src/main/java/redis/clients/jedis/csc/CacheEntry.java

```diff
@@ -1,30 +1,58 @@
 package redis.clients.jedis.csc;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.lang.ref.WeakReference;
+
 import redis.clients.jedis.Connection;
 import redis.clients.jedis.annots.Internal;
+import redis.clients.jedis.exceptions.JedisCacheException;
 
 @Internal
 public class CacheEntry<T> {
 
   private final CacheKey<T> cacheKey;
-  private final T value;
-  private final Connection connection;
+  private final WeakReference<Connection> connection;
+  private final byte[] bytes;
 
-  public CacheEntry(CacheKey<T> cacheKey, T value, Connection connection) {
+  public CacheEntry(CacheKey<T> cacheKey, T value, WeakReference<Connection> connection) {
     this.cacheKey = cacheKey;
-    this.value = value;
     this.connection = connection;
+    this.bytes = toBytes(value);
   }
 
   public CacheKey<T> getCacheKey() {
     return cacheKey;
   }
 
   public T getValue() {
-    return value;
+    return toObject(bytes);
   }
 
-  public Connection getConnection() {
+  public WeakReference<Connection> getConnection() {
     return connection;
   }
+
+  private static byte[] toBytes(Object object) {
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ObjectOutputStream oos = new ObjectOutputStream(baos)) {
+      oos.writeObject(object);
+      oos.flush();
+      oos.close();
+      return baos.toByteArray();
+    } catch (Exception e) {
+      throw new JedisCacheException("Failed to serialize object", e);
```

Comment:
I am wondering - should cache serialization fail the app?

Assuming there is some value that breaks serialization for some reason (temporary or not) - we could probably continue the normal program execution without disturbing the flow, only log the error in the log, and ignore the cache. In any case falling back to reading from the server should be completely safe, correct?

Reply:
Btw do we need to catch `Exception` here, could it be some more specific implementation?

Reply:
i ll give it a thought,,  a bit concerned about not caching when it is the expectation. 
imho warnings are tend to be out of sight when there are no bad things under investigation.

Reply:
About the catch-everything pattern (where you catch for the `Exception` or even worse `Throwable` level) is that of handling what you caught. Some exceptions could not be handled in a meaningful way (most `RuntimeException`s e.g. `NullPointerException`, `ClassCastException`) while others could. When you wrap all of them all in a JedisCacheException the users of the driver could assume they could catch it and handle it in some way (e.g. just ignore them) and could - involuntary - handle more than they expected (and hide some bug in the driver).

IMHO a good pattern would be to let all of these exceptions propagate naturally and only wrap checked exceptions. This would make our code - more or less - in line with the "unchecked exceptions only" club (you probably know there are two conflicting philosophies on checked vs unchecked exceptions).

Reply:
how about wrapping all into a custom cacheserialization and cachedeserialization exceptions. In case of deserialization we might choose to remove the entry from cache as an additional action. not sure about it.
for serialization fail, i am definitely in favor of throwing it regardless of logging it or not.

Reply:
Would not solve the catch-everything scenario I mentioned.
Anyway, since the JedisCacheException is ultimately a RuntimeException we don't have much of a problem, as most of the users are unlikely going to catch it explicitly.
Let's leave it like this.